File: Python/householder_qr_solver.py
import numpy as np

# Set seed for reproducibility
np.random.seed(0)

# (a) Generate A with known QR factorization
R = np.triu(np.random.randn(50, 50))
Q, _ = np.linalg.qr(np.random.randn(50, 50))
A = Q @ R

# (b) Compute QR factorization of A using Householder (numpy.linalg.qr)
Q2, R2 = np.linalg.qr(A)

# Compute backward error and orthogonality error
backward_error = np.linalg.norm(Q2 @ R2 - A, 2) / np.linalg.norm(A, 2)
orthogonality_error = np.linalg.norm(Q2.T @ Q2 - np.eye(50), 2)

print("Relative backward error ||Q2*R2 - A|| / ||A||  =", backward_error)
print("Orthogonality error ||Q2^T*Q2 - I||             =", orthogonality_error)

File: Matlab/householder_qr_solver.m
% Set the random seed for reproducibility
rng(0);

% (a) Generate A with known QR factorization
R = triu(randn(50));
[Q, ~] = qr(randn(50));
A = Q * R;

% (b) Compute QR factorization of A
[Q2, R2] = qr(A);

% Compute backward error
backward_error = norm(Q2 * R2 - A, 2) / norm(A, 2);

% Compute orthogonality error
orthogonality_error = norm(Q2' * Q2 - eye(50), 2);

% Display results
fprintf('Relative backward error ||Q2*R2 - A|| / ||A||  = %.2e\n', backward_error);
fprintf('Orthogonality error ||Q2^T*Q2 - I||             = %.2e\n', orthogonality_error);
